docker/dev_env/run.sh

```diff
@@ -2,10 +2,10 @@
 
 set -e
 
-container_name="openverse-dev-env"
+container_name="openverse-dev_env"
```

Comment:
Wait, why did this get changed _away_ from the hyphen now?

Reply:
Our style is hyphens between "openverse" and service name and underscores within the service name.

Reply:
This is going to force everyone using `ov` to rebuild their container from scratch, including anything cached in `opt` (like pre-commit environments).

Is this really necessary? And is the use of underscores in the app name a prescriptive "must do no matter what"? (I personally find the mixed usage ugly, but just my opinion)

Mostly just concerned about the effect on every one currently using `ov`.

Reply:
The only reason this PR exists is because I'm obsessing over a minor, largely inconsequential detail. I don't mind closing it if the consensus is that it's too much of a hassle. This was nagging at me since I noticed it and I had to make the PR just for closure. Nonetheless, I'll just give a bit of history that explains how we got here to this ugly mix of separators.

When we were using Docker Compose v1, the separator between the directory/repo/project name and the service name (hardcoded by Docker Compose) was `_`. So for consistency, we used service names with underscore separators (like "ingestion_server" or "upstream_db"). 

With Docker Compose v2, the separator was changed to `-`. By that time, I personally felt it was better to use a different separator with the "openverse" prefix so that we can conveniently split at `-` to separate the prefix.

Reply:
Gotcha, thanks for the background.

> so that we can conveniently split at - to separate the prefix

I suppose `image_name.replace("openverse-", "")` is either just as or more reliable than `image_name.split("-")[-1]`. If we named an image without the hypen/underscore encoding in mind, the split version fails every time. The replace (or even better, substring, `image_name.substring("openverse-".length)` in js/`image_name[len("openverse-"):]` in python) always works.

Well, anyway, I just wanted to point out that this will cause everyone to need to rebuild the dev image (which may be cached? In which case are the explicit image names necessary?), but in particular any state that exists _after_ the dev image is built during the course of usage (and in particular the pre-commit envs, because they do take a long time to spin up unfortunately), will introduce some tedium.

At the very least it's something to be contientious of and maybe let the team know so that anyone who doesn't notice this PR or see the image name difference doesn't end up frustrated and confused why their environment is suddenly slow again while it initialises the first time around.

Overall my vote is just going to be: whatever is the best thing to have long term. If the mix of separators is the best thing, fine.

Looking at the documentation about image names, I guess we could use `openverse/<image name>` too, and that would actually fit the specific usage more closely. From [the Docker docs on image names](https://docs.docker.com/reference/cli/docker/image/tag/#extended-description):

>     `[NAMESPACE/]REPOSITORY`: The first, optional component is typically a user's or an organization's namespace. The second, mandatory component is the repository name. When the namespace is not present, Docker uses library as the default namespace.

If we're going to cause an upheaval, and there's a "more correct" or obvious or widely shared approach to this kind of thing, we could go ahead and take the opportunity to do it now, rather than multiple times. I'm personally a lot more likely to remember the nuance between `openverse/` then `openverse-`.

Just my two cents! If we're delving into the weeds of little details, might as well delve all the way in and ask if there's a "perfect" version. Anecdotally, I can't recall any public docker images I've used that use underscore separators _at all_. Hyphens seem much more common :thinking:

But it's just little details. I agree consistency would be best above anything else. Consistency _and_ matching the exact features of whatever implementation we're relying on? Maybe even better...

While we're at it, for the images that are published, should we use the GHCR reference for them instead? Maybe then pulling would actually _be_ a benefit. An individual might get some stages of the build cached on that first run of the stack?

But it would still be nice to move those towards the same namespace approach of `openverse/`, if we can do that? I can't remember of we control https://github.com/openverse/, but GHCR didn't mind when I pushed an image with multiple namespaces, so `wordpress/openverse/<image name>` would work (example here: https://github.com/users/sarayourfriend/packages/container/package/openverse%2Ffrontend).

The OCI spec explicitly allows multiple namespaces: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pulling-manifests. Docker's documentation mentions that most repositories only support at most two slashes, but we use GHCR and it handles them fine. If we ever had to change registries, we can make sure we only change to a spec-compliant registry (benefits all around there, I'd say).

Look, bottom line for me is: let's pick whatever is the best thing we actually _want_, especially if this will be somewhat disruptive anyway, so that we aren't left groaning or wanting to make another potentially disruptive change again in the future. If that's the current overloaded meaning of hyphens and underscores: fine, all good, let's go for it, and let's do some work to make sure we document and maybe lint for that. If it's the namespaces, same deal, document and educate/lint.

Whichever way we go is fine with me. Like the slashes because I think they look nice and are easy to remember what they mean, but I ultimately don't care about the specific thing we choose here, as I agree the more fundamentally important thing is consistency. But if we have an opportunity to change towards a better approach (if it is indeed better) then we can take that now, if we want to.

Reply:
I have no strong preferences either.

1. I like your ideas with slashes and namespaces, but that seems like a pretty big change across our code and infrastructure and docs. I also see there is some uncertainty around support for the multiple namespaces spec across repositories (the link in your example does not work).
2. This PR, my second preference, involves 2 changes and that too only to the local dev environment. Everything else remains unchanged, and I find that stability quite nice. Also I only made the PR to resolve small discrepancies, I did not expect to be making huge changes.

I don't know how one could lint for image names but regardless of the choice, it could be quite easy to document the naming scheme somewhere.

Reply:
Sorry, I'm making a much of nothing in this thread. To clarify I am also not blocking this change (I would have requested changes if that was the case). The main important thing I want to stress is the disruption to `ov`, and to request that the disruption (and reason for it) is communicated clearly using our regular communication tools. Knowing a similar disruption won't happen again any time soon would also be good. So, if this hyphen/underscore meaning is meaningful enough for such a disruption, then yes, that should also be documented and made visible, which isn't inherent in just documenting. Linting image names to enforce an expectation could be done rather trivially using the output from `ov just dc config`. It doesn't cover instances like `ov` that are outside of the compose stack (but then I don't personally see why this image name needs to match anyway, as it is outside the application stack where any automation could reasonably be applied, but I'm not arguing one way or another here... like I said, I would have requested changes if I felt so strongly).

To me it is also worth questioning whether the hyphen/underscore meaning is valuable or reliable, if either of those are the goal, or what the consistency brings. This looks like just a trivial change in names but becuase of the effect on breaking the `ov` volume, it's unfortunately more than just a change in names :shrug:

For what it's worth, as far as I can tell the GHCR link does work (and I can `docker pull ghcr.io/sarayourfriend/openverse/frontend:testing`). Maybe you are referencing a different link?

Reply:
The link doesn't work for me.

https://github.com/WordPress/openverse/assets/16580576/6e6b30fd-78b6-4d16-8f84-4c1e5a585ef6

It might be a permissions issue.

```
❯ docker pull ghcr.io/sarayourfriend/openverse/frontend:testing
Error response from daemon: Head "https://ghcr.io/v2/sarayourfriend/openverse/frontend/manifests/testing": unauthorized
```

I feel like once this change is merged, the image names will be consistent enough that I might be satisfied and we could reliably keep them the same for a long time.

Also the effect on the `ov` volume is just that all the stuff will automatically be rebuilt. There will be a small increase in first run time post-merge, but mostly the workflow will remain unbroken and unchanged. I don't see this as a breaking change, and FWIW, we should be open to breaking changes with `ov` for a short while as it is fairly new, has not been widely adopted, and we're ironing out the kinks.

Reply:
Ah, it must be a permissions issue, yes.

> Also the effect on the ov volume is just that all the stuff will automatically be rebuilt. There will be a small increase in first run time post-merge, but mostly the workflow will remain unbroken and unchanged. I don't see this as a breaking change, and FWIW, we should be open to breaking changes with ov for a short while as it is fairly new, has not been widely adopted, and we're ironing out the kinks.

It's the pre-commit environments that take the longest for me, at least, and it's a pain and extremely disruptive because it's always just as I am usually wrapping up a bit of work. The renovate config validator takes so long! Maybe it shouldn't be in the local pre-commit, and we could just validate it in CI.

I agree we can accept breaking changes for a while, but I also want to be mindful of the fact that because it's new and different, it can feel extra frustrating if things suddenly feel inconsistent or random. Like, if someone doesn't know about this PR and suddenly they have to rebuild the pre-commit env again just before merging, it will be frustrating, confusing, and seem like `ov` has introduced an inscruitable and frustrating lack of transparency and consistency into the development environment. That's why it's important to communicate the change and why it happened. I know I would be frustrated and sceptical if `ov` suddenly seemed to just make things more confusing and slower when the other way worked fine (for me), and I'd question it's value if I didn't have the context of why e.g., committing suddenly was taking a very long time because of the pre-commit environment creation happening _again_.

As I said:

> This looks like just a trivial change in names but becuase of the effect on breaking the ov volume, it's unfortunately more than just a change in names 🤷

I think that's mischaracterising it, and a small increase in the first run time is also mischaracterising it, ignoring a significant frustration. Maybe I'm the only one who finds a slow pre-commit a pain in the butt, but I really do find it almost intolerable (I almost always `--no-verify` the frontend's pre-push hooks, which are truly intolerably slow to me, and often run on rebased pushes). Anyway, I just don't want that to be brushed off and because `ov` is new it's good to make sure everyone knows about it.

Will you communicate the change, the reason for it, and the effect of it to everyone? In Slack, at the very least? I think the PR description should also include a note indicating that. I just don't think it's as trivial a change for that first touch as you've made it out to be.

Reply:
> It's the pre-commit environments that take the longest for me

That is true. pre-commit setup is a bit slow and any hooks that take a long time to be configured or run should be relegated to the CI workflow. In fact @AetherUnbound and I were discussing the making pre-commit hooks opt-in, so that there is one fewer barrier for new contributors.

> That's why it's important to communicate the change and why it happened.

That is definitely something that we will do. Once this PR is merged, we should post a message on Slack about the change and what actions a contributor needs to take to have everything working as before.

> I almost always --no-verify the frontend's pre-push hooks

I feel like `--no-verify` carries a negative stigma which it does not deserve. Since we have a fast-failing lint job in CI, we should feel free to use `--no-verify` if that speeds things up or reduces frustration.

> Will you communicate the change, the reason for it, and the effect of it to everyone?

Yes!